runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/IsolationChannel.java

```diff
@@ -0,0 +1,246 @@
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ManagedChannel;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.Metadata;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.MethodDescriptor;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.Status;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+
+/**
+ * A {@link ManagedChannel} that creates a dynamic # of cached channels to the same endpoint such
+ * that each active rpc has its own channel.
+ */
+@Internal
+class IsolationChannel extends ManagedChannel {
+  static final Logger LOG = Logger.getLogger(IsolationChannel.class.getName());
+
+  private final Supplier<ManagedChannel> channelFactory;
+
+  @GuardedBy("this")
+  private final List<ManagedChannel> channelCache;
+
+  @GuardedBy("this")
+  private final Set<ManagedChannel> usedChannels;
+
+  @GuardedBy("this")
+  private boolean shutdownStarted = false;
+
+  private final String authority;
+
+  // Expected that supplier returns channels to the same endpoint with the same authority.
+  public static IsolationChannel create(Supplier<ManagedChannel> channelFactory) {
+    return new IsolationChannel(channelFactory);
+  }
+
+  @VisibleForTesting
+  IsolationChannel(Supplier<ManagedChannel> channelFactory) {
+    this.channelFactory = channelFactory;
+    this.channelCache = new ArrayList<>();
+    ManagedChannel channel = channelFactory.get();
+    this.authority = channel.authority();
+    this.channelCache.add(channel);
+    this.usedChannels = new HashSet<>();
+  }
+
+  @Override
+  public String authority() {
+    return authority;
+  }
+
+  /** Pick an unused channel from the set or create one and then create a {@link ClientCall}. */
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(
+      MethodDescriptor<ReqT, RespT> methodDescriptor, CallOptions callOptions) {
+    ManagedChannel channel = getChannel();
+    return new ReleasingClientCall<>(this, channel, channel.newCall(methodDescriptor, callOptions));
+  }
+
+  private synchronized ManagedChannel getChannel() {
+    if (!channelCache.isEmpty()) {
+      ManagedChannel result = channelCache.remove(channelCache.size() - 1);
+      usedChannels.add(result);
+      return result;
+    }
+    if (shutdownStarted) {
+      // Avoid creating a new channel if we're shutting down and just use an existing one
+      // that has already started shutting down.
+      Preconditions.checkArgument(!usedChannels.isEmpty());
+      return usedChannels.iterator().next();
+    }
+
+    ManagedChannel result = channelFactory.get();
+    usedChannels.add(result);
+    return result;
+  }
+
+  @Override
+  public synchronized ManagedChannel shutdown() {
```

Comment:
can we remove the synchronized from method signature and add a synchronized(this) block inside? That'll allow changing internal locking without changing the method signature. With synchronized in the signature, other classes might start relying on the this locking behavior.

Reply:
Done